man/man3/Post.3
.TH "Post" 3 "Fri Oct 7 2011" "Version 1.0" "Launchpad" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Post \- 
.SH SYNOPSIS
.br
.PP
.PP
Inherits \fBContent\fP.
.PP
Inherited by \fBPage\fP.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBis_in_category\fP ($category)"
.br
.ti -1c
.RI "\fBcategory_name\fP ()"
.br
.ti -1c
.RI "\fBget_next_post\fP ($category=NULL)"
.br
.ti -1c
.RI "\fBget_previous_post\fP ($category=NULL)"
.br
.ti -1c
.RI "\fBget_adjacent_post\fP ($category=NULL, $previous=false)"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBfind_by_category_id\fP ($id, $order='created DESC', $limit=NULL, $offset=NULL, $return_count=false)"
.br
.ti -1c
.RI "static \fBfind_by_category_title\fP ($titles, $order='created DESC', $limit=NULL, $offset=NULL, $return_count=false)"
.br
.ti -1c
.RI "static \fBfind_except_in_categories\fP ($categories, $order='created DESC', $limit=NULL, $offset=NULL, $return_count=false)"
.br
.ti -1c
.RI "static \fBfind_by_tag\fP ($tag, $order='created DESC', $limit=NULL, $offset=NULL, $return_count=false)"
.br
.ti -1c
.RI "static \fBfind_by_dates\fP ($month, $day=NULL, $year=NULL, $categories=NULL, $order='created DESC', $limit=NULL, $offset=NULL, $return_count=false)"
.br
.ti -1c
.RI "static \fBget_dates\fP ($month=NULL, $year=NULL, $categories=NULL)"
.br
.ti -1c
.RI "static \fBget_authors\fP ($categories)"
.br
.ti -1c
.RI "static \fBget_categories\fP ()"
.br
.in -1c
.SH "Member Function Documentation"
.PP 
.SS "Post::category_name ()"Retuns the current category name 
.PP
\fBReturns:\fP
.RS 4
string 
.RE
.PP

.SS "static Post::find_by_category_id ($id, $order = \fC'created DESC'\fP, $limit = \fCNULL\fP, $offset = \fCNULL\fP, $return_count = \fCfalse\fP)\fC [static]\fP"Returns an array of content by
.PP
\fBParameters:\fP
.RS 4
\fIint\fP id - category id 
.br
\fI$order\fP = 'created DESC' - The field name by which to oder the results. Use DESC or ASC to swap order 
.br
\fI$limit\fP int - limit the amount of content to retrieve 
.br
\fI$offset\fP int - offset of the first row to return 
.br
\fI$return_count\fP boolen - If true will return a row count instead of an array of \fBPost\fP object. Usefull for doing pagination. 
.RE
.PP
\fBReturns:\fP
.RS 4
array 
.RE
.PP

.SS "static Post::find_by_category_title ($titles, $order = \fC'created DESC'\fP, $limit = \fCNULL\fP, $offset = \fCNULL\fP, $return_count = \fCfalse\fP)\fC [static]\fP"Returns an array of Posts by category name
.PP
\fBParameters:\fP
.RS 4
\fIstring\fP | array $titles - The \fBCategory\fP title as a string or an array of strings 
.br
\fI$order\fP = 'created DESC' - The field name by which to oder the results. Use DESC or ASC to swap order 
.br
\fI$limit\fP int - limit the amount of content to retrieve 
.br
\fI$offset\fP int - offset of the first row to return 
.br
\fI$return_count\fP boolen - If true will return a row count instead of an array of \fBPost\fP object. Usefull for doing pagination. 
.RE
.PP
\fBReturns:\fP
.RS 4
array 
.RE
.PP

.SS "static Post::find_by_dates ($month, $day = \fCNULL\fP, $year = \fCNULL\fP, $categories = \fCNULL\fP, $order = \fC'created DESC'\fP, $limit = \fCNULL\fP, $offset = \fCNULL\fP, $return_count = \fCfalse\fP)\fC [static]\fP"Retrives all the post created on a specific month day year.
.PP
\fBParameters:\fP
.RS 4
\fI$month\fP - The Post's month it was created in 
.br
\fI$day,=NULL\fP - The Post's day it was created in 
.br
\fI$year,=NULL\fP - The Post's year it was created in 
.br
\fIarray\fP | string $categories=NULL - Only retive posts of specific categories 
.br
\fI$order\fP = 'created DESC' - The field name by which to order the results. Use DESC or ASC to swap order 
.br
\fI$limit\fP - limit the amount of content to retrieve 
.br
\fI$offset\fP - offset of the first row to return 
.br
\fI$return_count\fP boolen - If true will return a row count instead of an array of \fBPost\fP object. Usefull for doing pagination. 
.RE
.PP
\fBReturns:\fP
.RS 4
array - Array containg result Posts 
.RE
.PP

.SS "static Post::find_by_tag ($tag, $order = \fC'created DESC'\fP, $limit = \fCNULL\fP, $offset = \fCNULL\fP, $return_count = \fCfalse\fP)\fC [static]\fP"Returns an array content linked to specified tag
.PP
\fBParameters:\fP
.RS 4
\fIstring\fP | integer $tag - A \fBTag\fP id or tag name 
.br
\fI$order\fP = 'created DESC' - The field name by which to order the results. Use DESC or ASC to swap order 
.br
\fI$limit\fP - limit the amount of content to retrieve 
.br
\fI$offset\fP - offset of the first row to return 
.br
\fI$return_count\fP boolen - If true will return a row count instead of an array of \fBPost\fP object. Usefull for doing pagination. 
.RE
.PP
\fBReturns:\fP
.RS 4
array 
.RE
.PP

.SS "static Post::find_except_in_categories ($categories, $order = \fC'created DESC'\fP, $limit = \fCNULL\fP, $offset = \fCNULL\fP, $return_count = \fCfalse\fP)\fC [static]\fP"Returns an array content excluding provided categories ids
.PP
\fBParameters:\fP
.RS 4
\fIcategories\fP array - categories to exclude 
.br
\fI$order\fP = 'created DESC' - The field name by which to order the results. Use DESC or ASC to swap order 
.br
\fI$limit\fP - limit the amount of content to retrieve 
.br
\fI$offset\fP - offset of the first row to return 
.br
\fI$return_count\fP boolen - If true will return a row count instead of an array of \fBPost\fP object. Usefull for doing pagination. 
.RE
.PP
\fBReturns:\fP
.RS 4
array 
.RE
.PP

.SS "Post::get_adjacent_post ($category = \fCNULL\fP, $previous = \fCfalse\fP)"Returns the an adjacently created post
.PP
\fBParameters:\fP
.RS 4
\fIstring\fP | number $category = NULL - restrict posts to specified category 
.br
\fI$previous\fP = true - true will return the next post false will return the previous 
.RE
.PP
\fBReturns:\fP
.RS 4
\fBPost\fP 
.RE
.PP

.SS "static Post::get_authors ($categories)\fC [static]\fP"Returns all the usres who have publish Posts
.PP
\fBParameters:\fP
.RS 4
\fIarray\fP | string - Specify a specific category of Posts; 
.RE
.PP
\fBReturns:\fP
.RS 4
array - Returns and array of \fBUser\fP objects 
.RE
.PP

.SS "static Post::get_categories ()\fC [static]\fP"Returns all categories
.PP
\fBReturns:\fP
.RS 4
array 
.RE
.PP

.SS "static Post::get_dates ($month = \fCNULL\fP, $year = \fCNULL\fP, $categories = \fCNULL\fP)\fC [static]\fP"Retrives all the months and or years Posts have been created. This is useful when created navigation based on dates;
.PP
\fBParameters:\fP
.RS 4
\fI$month\fP - The month to search 
.br
\fI$year,=NULL\fP - The year to search 
.br
\fIarray\fP | string $categories=NULL - Only search Posts of specific categories 
.RE
.PP
\fBReturns:\fP
.RS 4
array - An array of months all posts were created 
.RE
.PP

.SS "Post::get_next_post ($category = \fCNULL\fP)"Returns the next created post
.PP
\fBParameters:\fP
.RS 4
\fIstring\fP | number $category = NULL - restrict posts to specified category 
.RE
.PP
\fBReturns:\fP
.RS 4
\fBPost\fP 
.RE
.PP

.SS "Post::get_previous_post ($category = \fCNULL\fP)"Returns the previoulsly created post
.PP
\fBParameters:\fP
.RS 4
\fIstring\fP | number $category = NULL - restrict posts to specified category 
.RE
.PP
\fBReturns:\fP
.RS 4
\fBPost\fP 
.RE
.PP

.SS "Post::is_in_category ($category)"Checks to see if the post is in a \fBCategory\fP
.PP
\fBParameters:\fP
.RS 4
\fIint\fP | string $category - The category 
.RE
.PP
\fBReturns:\fP
.RS 4
boolean 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for Launchpad from the source code.


lib/post.class.php
<?php
	/**
	* 
	*/
	class Post extends Content
	{										
		
		/**
		 * Checks to see if the post is in a Category
		 *
		 * @param int | string $category - The category
		 * @return boolean
		 **/
		public function is_in_category($category)
		{
			if(empty($this->category_id)){
				return false;
			}else{
				if(is_numeric($category)){
					return $category == $this->category_id;
				}elseif( is_string ($category) ){
					$sql = "SELECT id FROM categories WHERE title = '$category'";				
					$result = Category::find_by_sql($sql);
					return ($result > 0 && $result[0]->id == $this->category_id) ? true : false;
				}
			
			}
		}

		/**
		 * Retuns the current category name
		 * @return string
		 **/
		public function category_name(){
			$cat = Category::find_by_id($this->category_id);
			return $cat->title;	
		}
		/**
		 * Returns the next created post
		 *
		 * @param string | number $category = NULL - restrict posts to specified category
		 * @return Post
		 **/
		public function get_next_post($category=NULL)
		{
			return $this->get_adjacent_post($category, false);
		}
		/**
		 * Returns the previoulsly created post
		 *
		 * @param string | number $category = NULL - restrict posts to specified category
		 * @return Post
		 **/
		public function get_previous_post($category=NULL)
		{
			return $this->get_adjacent_post($category, true);
		}
		/**
		 * Returns the an adjacently created post
		 *
		 * @param string | number $category = NULL - restrict posts to specified category
		 * @param boolean $previous = true - true will return the next post false will return the previous
		 * @return Post
		 **/
		public function get_adjacent_post($category=NULL, $previous=false )
		{
			$created_str = ($previous) ? " AND created > '$this->created'" : " AND created < '$this->created'";
			
			$sql = "SELECT * FROM content";
			$sql .= " WHERE type ='".ContentType::POST."'";
			$sql .= " AND status ='".ContentStatus::PUBLISHED."'";
			$sql .= $created_str;
			if( isset($category) ){
				if(is_numeric($category)){
					$sql .= " AND category_id = $category";
				}elseif( is_string ($category) ){
					$sql .= " AND category_id = (SELECT id FROM categories WHERE title = '$category')";				
				}
			}
			$sql .= " ORDER BY created";
			$sql .= ($previous) ? ' ASC' : ' DESC'; 
			$sql .= " LIMIT 1";
			$result = Post::find_by_sql($sql);
			if( empty($result)) {
				$sql=str_replace($created_str, '', $sql);
				$result = Post::find_by_sql($sql);
				return !empty($result) ? array_pop($result) : false;
			}else{
                return array_pop($result);
			}
		}
		/**
		* Returns an array of content by
		*
		* @param int id - category id
		* @param string $order = "created DESC" - The field name by which to oder the results. Use DESC or ASC to swap order
		* @param $limit int - limit the amount of content to retrieve
		* @param $offset int - offset of the first row to return
		* @param $return_count boolen - If true will return a row count instead of an array of Post object. Usefull for doing pagination.
		* @return array
		**/
		public static function find_by_category_id($id, $order="created DESC", $limit=NULL, $offset=NULL, $return_count=false)
		{
			$select = ($return_count) ? "COUNT(*) as count": "*";
			$sql = "SELECT $select FROM content";
			$sql .= " WHERE type ='".ContentType::POST."'";
			$sql .= " AND status ='".ContentStatus::PUBLISHED."'";
			$sql .= " AND category_id={$id}";
			if(isset($order)){
				$sql .= " ORDER BY ".$order;
			}
			if(isset($offset) && isset($limit)){
				$sql .= " LIMIT ".$offset.", ".$limit;
			}elseif(isset($limit)){
				$sql .= " LIMIT ".$limit;
			}
			$result = Post::find_by_sql($sql);
			if($return_count) $result = $result[0]->count;
			return !empty($result) ? $result : false;
		}
		/**
		* Returns an array of Posts by category name
		*
		* @param string | array $titles - The Category title as a string or an array of strings
		* @param string $order = "created DESC" - The field name by which to oder the results. Use DESC or ASC to swap order
		* @param $limit int - limit the amount of content to retrieve
		* @param $offset int - offset of the first row to return
		* @param $return_count boolen - If true will return a row count instead of an array of Post object. Usefull for doing pagination.
		* @return array
		**/
		public static function find_by_category_title($titles, $order="created DESC", $limit=NULL, $offset=NULL, $return_count=false)
		{
			$select = ($return_count)?"COUNT(*) as count": "content.".implode(", content.", parent::$db_fields);
			$sql = "SELECT $select FROM content, categories";
			$sql .= " WHERE content.type ='".ContentType::POST."'";
			$sql .= " AND content.status ='".ContentStatus::PUBLISHED."'";
			$sql .= " AND content.category_id=categories.id";
			if(is_string($titles)){
				$sql .= " AND categories.title='$titles'";
			} elseif( is_array($titles) ) {
				$sql .= " AND categories.title IN ('".implode( "','" , $titles )."')";
			}
			if(isset($order)){
				$sql .= " ORDER BY ".$order;
			}
			if(isset($offset) && isset($limit)){
				$sql .= " LIMIT ".$offset.", ".$limit;
			}elseif(isset($limit)){
				$sql .= " LIMIT ".$limit;
			}
			$result = Post::find_by_sql($sql);
			if($return_count) $result = $result[0]->count;
			return !empty($result) ? $result : false;
		}
		
		/**
		* Returns an array content excluding provided categories ids
		*
		* @param categories array - categories to exclude
		* @param string $order = "created DESC" - The field name by which to order the results. Use DESC or ASC to swap order
		* @param integer $limit - limit the amount of content to retrieve
		* @param integer $offset - offset of the first row to return
		* @param $return_count boolen - If true will return a row count instead of an array of Post object. Usefull for doing pagination.
		* @return array
		**/
		public static function find_except_in_categories( $categories, $order='created DESC', $limit=NULL, $offset=NULL, $return_count=false)
		{
			$select = ($return_count) ? "COUNT(*) as count": "*";
			$sql = "SELECT $select FROM content";
			$sql .= " WHERE type ='".ContentType::POST."'";
			$sql .= ' AND category_id NOT IN ('.implode("," , $categories ).")";
			$sql .= " AND status ='".ContentStatus::PUBLISHED."'";
			if(isset($order)){
				$sql .= " ORDER BY ".$order;
			}
			if(isset($offset) && isset($limit)){
				$sql .= " LIMIT ".$offset.", ".$limit;
			}elseif(isset($limit)){
				$sql .= " LIMIT ".$limit;
			}
			
			$result = Post::find_by_sql($sql);
			if($return_count) $result = $result[0]->count;
			return !empty($result) ? $result : false;
		}				
		
		/**
		* Returns an array content linked to specified tag
		*
		* @param string | integer $tag - A Tag id or tag name
		* @param string $order = "created DESC" - The field name by which to order the results. Use DESC or ASC to swap order
		* @param integer $limit - limit the amount of content to retrieve
		* @param integer $offset - offset of the first row to return
		* @param $return_count boolen - If true will return a row count instead of an array of Post object. Usefull for doing pagination.
		* @return array
		**/
		public static function find_by_tag($tag, $order='created DESC', $limit=NULL, $offset=NULL, $return_count=false ) 
		{
			$select = ($return_count) ? "COUNT(*) as count": "*";
			$sql = "SELECT $select FROM content";
			$sql .= " JOIN content_x_tags ON content.id = content_x_tags.content_id";
			$sql .= " JOIN tags ON tags.id = content_x_tags.tag_id";
			if( is_numeric($tag) ){
				$sql .= " WHERE tags.id = $tag";
			}elseif( is_string($tag) ){
				$sql .= " WHERE tags.tag = '$tag'";
			}
			$sql .= " AND status='". ContentStatus::PUBLISHED."'"; 
			$sql .= " AND type='".ContentType::POST."'";
			if(isset($order)){
				$sql .= " ORDER BY ".$order;
			}
			if(isset($offset) && isset($limit)){
				$sql .= " LIMIT ".$offset.", ".$limit;
			}elseif(isset($limit)){
				$sql .= " LIMIT ".$limit;
			}
			$result= Post::find_by_sql($sql);
			if($return_count) $result = $result[0]->count;
			return !empty($result) ? $result : false ;
		}

		/**
		 * Retrives all the post created on a specific month day year. 
		 *
		 * @param integer $month - The Post's month it was created in
		 * @param integer $day=NULL - The Post's day it was created in
		 * @param integer $year=NULL - The Post's year it was created in
		 * @param array | string $categories=NULL - Only retive posts of specific categories
		 * @param string $order = "created DESC" - The field name by which to order the results. Use DESC or ASC to swap order
		 * @param integer $limit - limit the amount of content to retrieve
		 * @param integer $offset - offset of the first row to return
		 * @param $return_count boolen - If true will return a row count instead of an array of Post object. Usefull for doing pagination.
		 * @return array - Array containg result Posts
		 **/
		public static function find_by_dates($month, $day=NULL, $year=NULL, $categories=NULL, $order='created DESC', $limit=NULL, $offset=NULL, $return_count=false)
		{
			$date = mktime(0, 0, 0, $month); 
			if (isset($year)){
			   $date = ( isset($day) ) ? mktime(0, 0, 0, $month, $day, $year) : mktime(0, 0, 0, $month, 1, $year); 
			};
			$date = date("Y-m-d H:i:s", $date);
			$select = ($return_count) ? "COUNT(*) as count": "content.".implode(", content.", parent::$db_fields).", categories.id";
			$sql = "SELECT $select FROM content, categories WHERE content.created BETWEEN '".$date."' AND LAST_DAY('".$date."')";
			$sql .= " AND content.type ='".ContentType::POST."'";
			$sql .= " AND content.status ='".ContentStatus::PUBLISHED."'";
			if(isset($categories)){
				$sql .= " AND content.category_id = categories.id";
				if(is_string($categories)){
					$sql .= " AND categories.title='$categories'";
				} elseif( is_array($categories) ) {
					$sql .= " AND categories.title IN ('".implode( "','" , $categories )."')";
				}
			}
			if(isset($year)){
				$sql .= " AND YEAR(content.created) = $year";
			}                                                     
			if(isset($month)){
				$sql .= " AND MONTH(content.created) = $month";
			}                                                     
			if(isset($order)){
				$sql .= " ORDER BY ".$order;
			}
			if(isset($offset) && isset($limit)){
				$sql .= " LIMIT ".$offset.", ".$limit;
			}elseif(isset($limit)){
				$sql .= " LIMIT ".$limit;
			}

			$result = Post::find_by_sql($sql);
			if($return_count) $result = $result[0]->count;
			return !empty($result) ? $result : false;
		}

		/**
		 * Retrives all the months and or years Posts have been created. This is useful when created navigation based on dates;
		 *
		 * @param integer $month - The month to search
		 * @param integer $year=NULL - The year to search
		 * @param array | string $categories=NULL - Only search Posts of specific categories
		 * @return array - An array of months all posts were created
		 **/
		public static function get_dates($month=NULL, $year=NULL, $categories=NULL )
		{
			global $database;
			$sql="SELECT DAY(content.created) AS day, MONTH(content.created) AS month, YEAR(content.created) AS year";
			$sql .= " FROM content, categories";
			$sql .= " WHERE type ='".ContentType::POST."'";
			$sql .= " AND status ='".ContentStatus::PUBLISHED."'";
			if(isset($categories)){
				$sql .= " AND content.category_id = categories.id";
				if(is_string($categories)){
					$sql .= " AND categories.title='$categories'";
				} elseif( is_array($categories) ) {
					$sql .= " AND categories.title IN ('".implode( "','" , $categories )."')";
				}
			}
			if(isset($year)){
				$sql .= " AND YEAR(content.created) = $year";
			}                                                     
			if(isset($month)){
				$sql .= " AND MONTH(content.created) = $month";
			}                                                     
			$sql .= " GROUP BY month, year";
			$sql .= " ORDER BY created";
			$result = $database->query($sql);
			$result = $result->fetchAll(PDO::FETCH_ASSOC);
			return $result;
		}

		/**
		 * Returns all the usres who have publish Posts
		 *
		 * @param array | string  -  Specify a specific category of Posts;
		 * @return array -  Returns and array of User objects
		 **/
		public static function get_authors($categories)
		{
			$sql = "SELECT * FROM users WHERE users.id IN ( ";
			$sql .= "SELECT author_id FROM content";
			$sql .= " WHERE type ='".ContentType::POST."'";
			$sql .= " AND status ='".ContentStatus::PUBLISHED."'";
			if(isset($categories)){
				$sql .= " AND content.category_id = categories.id";
				if(is_string($categories)){
					$sql .= " AND categories.title='$categories'";
				} elseif( is_array($categories) ) {
					$sql .= " AND categories.title IN ('".implode( "','" , $categories )."')";
				}
			}
			$sql .= "GROUP BY author_id )";
			$result = User::find_by_sql($sql);
			return $result;
		}

		/**
		* Returns all categories
		*
		* @return array
		**/
		public static function get_categories()
		{
			$categories = Category::find_all();
			return !empty($categories) ? $categories : false;
		}


				
	}
	
?>
